File: web/src/stats.js
/** 과거 회차에서 번호별 통계를 뽑아낸다. */
import { NUMBER_POOL, profile, sortedNums } from './metrics.js';

/** 번호쌍을 Map 키로 쓰기 위한 문자열. 작은 번호가 앞. */
export const pairKey = (a, b) => (a < b ? `${a}-${b}` : `${b}-${a}`);

const inc = (map, key, by = 1) => map.set(key, (map.get(key) || 0) + by);

/** 회차 목록을 회차 번호 오름차순으로. */
const byDrawNo = draws => [...draws].sort((a, b) => a.no - b.no);

/**
 * 번호 1~45에 대한 출현 통계.
 * frequency/recent/gap/pairs/bonusFrequency 는 모두 Map.
 */
export function build(draws, recentWindow = 30) {
  const st = {
    drawsUsed: draws.length,
    recentWindow,
    frequency: new Map(),
    recent: new Map(),
    gap: new Map(),
    pairs: new Map(),
    bonusFrequency: new Map(),
  };
  if (!draws.length) {
    for (const n of NUMBER_POOL) st.gap.set(n, 0);
    return st;
  }

  const ordered = byDrawNo(draws);
  for (const draw of ordered) {
    for (const n of draw.numbers) inc(st.frequency, n);
    inc(st.bonusFrequency, draw.bonus);
    const nums = sortedNums(draw.numbers);
    for (let i = 0; i < nums.length; i++)
      for (let j = i + 1; j < nums.length; j++) inc(st.pairs, pairKey(nums[i], nums[j]));
  }

  for (const draw of ordered.slice(-recentWindow)) {
    for (const n of draw.numbers) inc(st.recent, n);
  }

  const lastSeen = new Map(NUMBER_POOL.map(n => [n, -1]));
  ordered.forEach((draw, idx) => {
    for (const n of draw.numbers) lastSeen.set(n, idx);
  });
  const total = ordered.length;
  for (const n of NUMBER_POOL) {
    const idx = lastSeen.get(n);
    st.gap.set(n, idx >= 0 ? total - 1 - idx : total);
  }
  return st;
}

export const meanFrequency = st => (st.drawsUsed ? (st.drawsUsed * 6) / 45 : 0);

/**
 * 파이썬 Counter.most_common(n) 과 같은 순서:
 * 횟수 내림차순, 동점이면 먼저 등록된 순(= 번호가 처음 나온 순).
 */
function mostCommon(counter, n) {
  const keys = [...counter.keys()];                       // Map 은 삽입 순서를 지킨다
  const order = new Map(keys.map((k, i) => [k, i]));
  return keys
    .sort((a, b) => (counter.get(b) - counter.get(a)) || (order.get(a) - order.get(b)))
    .slice(0, n)
    .map(k => [k, counter.get(k)]);
}

/** 최근 창에서 많이 나온 번호. */
export const hot = (st, n = 10) => mostCommon(st.recent, n);

/** 오래 안 나온 번호. 파이썬은 gap 내림차순 정렬(동점이면 번호 오름차순). */
export function cold(st, n = 10) {
  return [...st.gap.entries()]
    .sort((a, b) => (b[1] - a[1]) || (a[0] - b[0]))
    .slice(0, n);
}

/** 해당 번호와 같이 나온 적 많은 번호(궁합수). */
export function companions(st, number, n = 6) {
  return NUMBER_POOL.filter(other => other !== number)
    .map(other => [other, st.pairs.get(pairKey(number, other)) || 0])
    .sort((a, b) => (b[1] - a[1]) || (a[0] - b[0]))
    .slice(0, n);
}

/** 당첨조합들이 실제로 어떤 모양이었는지 요약 (= '대한민국 로또 평균치'). */
export function profileStats(draws) {
  const ordered = byDrawNo(draws);
  if (!ordered.length) throw new Error('분석할 회차 데이터가 없습니다.');

  const sums = [], ends = [];
  const oddD = new Map(), lowD = new Map(), acD = new Map(), carryD = new Map();
  let withConsecutive = 0;

  ordered.forEach((draw, idx) => {
    const prev = idx ? ordered[idx - 1].numbers : [];
    const p = profile(draw.numbers, prev);
    sums.push(p.total);
    ends.push(p.endSum);
    inc(oddD, p.odd);
    inc(lowD, p.low);
    inc(acD, p.ac);
    if (idx) inc(carryD, p.carryover);
    if (p.consecutive) withConsecutive++;
  });

  const orderedSums = [...sums].sort((a, b) => a - b);
  const lo = orderedSums[Math.floor(orderedSums.length * 0.10)];
  const hi = orderedSums[Math.min(orderedSums.length - 1, Math.floor(orderedSums.length * 0.90))];
  const mean = arr => arr.reduce((a, b) => a + b, 0) / arr.length;

  return {
    count: ordered.length,
    meanSum: mean(sums),
    sumRange80: [lo, hi],
    oddDistribution: oddD,
    lowDistribution: lowD,
    acDistribution: acD,
    endSumMean: mean(ends),
    consecutiveRatio: withConsecutive / ordered.length,
    carryoverDistribution: carryD,
  };
}

File: lottoracle/stats.py
"""과거 회차에서 번호별 통계를 뽑아낸다."""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field
from itertools import combinations
from typing import Sequence

from .data import Draw
from .metrics import NUMBER_POOL, profile


@dataclass
class NumberStats:
    """번호 1~45에 대한 출현 통계."""

    draws_used: int = 0
    frequency: Counter = field(default_factory=Counter)      # 전체 출현 횟수
    recent: Counter = field(default_factory=Counter)         # 최근 N회 출현 횟수
    recent_window: int = 0
    gap: dict[int, int] = field(default_factory=dict)        # 미출현 회차 수
    pairs: Counter = field(default_factory=Counter)          # 동반 출현(궁합수)
    bonus_frequency: Counter = field(default_factory=Counter)

    @property
    def mean_frequency(self) -> float:
        return (self.draws_used * 6) / 45 if self.draws_used else 0.0

    def hot(self, n: int = 10) -> list[tuple[int, int]]:
        return self.recent.most_common(n)

    def cold(self, n: int = 10) -> list[tuple[int, int]]:
        return sorted(self.gap.items(), key=lambda kv: -kv[1])[:n]

    def companions(self, number: int, n: int = 6) -> list[tuple[int, int]]:
        """해당 번호와 같이 나온 적 많은 번호(궁합수)."""
        scored = [
            (other, self.pairs[frozenset((number, other))])
            for other in NUMBER_POOL
            if other != number
        ]
        return sorted(scored, key=lambda kv: -kv[1])[:n]


def build(draws: Sequence[Draw], recent_window: int = 30) -> NumberStats:
    st = NumberStats(draws_used=len(draws), recent_window=recent_window)
    if not draws:
        st.gap = {n: 0 for n in NUMBER_POOL}
        return st

    ordered = sorted(draws, key=lambda d: d.no)
    for draw in ordered:
        st.frequency.update(draw.numbers)
        st.bonus_frequency[draw.bonus] += 1
        for pair in combinations(sorted(draw.numbers), 2):
            st.pairs[frozenset(pair)] += 1

    for draw in ordered[-recent_window:]:
        st.recent.update(draw.numbers)

    last_seen = {n: -1 for n in NUMBER_POOL}
    for idx, draw in enumerate(ordered):
        for n in draw.numbers:
            last_seen[n] = idx
    total = len(ordered)
    st.gap = {n: (total - 1 - idx if idx >= 0 else total) for n, idx in last_seen.items()}
    return st


# ----------------------------------------------------------- 회차 전체의 평균치

@dataclass
class DrawProfileStats:
    """당첨조합들이 실제로 어떤 모양이었는지 요약 (= '대한민국 로또 평균치')."""

    count: int
    mean_sum: float
    sum_range_80: tuple[int, int]
    odd_distribution: Counter
    low_distribution: Counter
    ac_distribution: Counter
    end_sum_mean: float
    consecutive_ratio: float
    carryover_distribution: Counter

    def render(self) -> str:
        def dist(counter: Counter) -> str:
            total = sum(counter.values()) or 1
            return "  ".join(
                f"{k}:{v}회({v / total:.0%})" for k, v in sorted(counter.items())
            )

        return "\n".join(
            [
                f"분석 회차 수      : {self.count}",
                f"당첨번호 합 평균  : {self.mean_sum:.1f} (이론 기대값 138)",
                f"합계 80% 구간     : {self.sum_range_80[0]} ~ {self.sum_range_80[1]}",
                f"홀수 개수 분포    : {dist(self.odd_distribution)}",
                f"저구간(1~22) 분포 : {dist(self.low_distribution)}",
                f"AC값 분포         : {dist(self.ac_distribution)}",
                f"끝수합 평균       : {self.end_sum_mean:.1f}",
                f"연속수 포함 비율  : {self.consecutive_ratio:.1%}",
                f"이월수 분포       : {dist(self.carryover_distribution)}",
            ]
        )


def profile_stats(draws: Sequence[Draw]) -> DrawProfileStats:
    ordered = sorted(draws, key=lambda d: d.no)
    if not ordered:
        raise ValueError("분석할 회차 데이터가 없습니다.")

    sums, ends = [], []
    odd_d, low_d, ac_d, carry_d = Counter(), Counter(), Counter(), Counter()
    with_consecutive = 0

    for idx, draw in enumerate(ordered):
        prev = ordered[idx - 1].numbers if idx else ()
        p = profile(draw.numbers, prev)
        sums.append(p.total)
        ends.append(p.end_sum)
        odd_d[p.odd] += 1
        low_d[p.low] += 1
        ac_d[p.ac] += 1
        if idx:
            carry_d[p.carryover] += 1
        if p.consecutive:
            with_consecutive += 1

    ordered_sums = sorted(sums)
    lo = ordered_sums[int(len(ordered_sums) * 0.10)]
    hi = ordered_sums[min(len(ordered_sums) - 1, int(len(ordered_sums) * 0.90))]

    return DrawProfileStats(
        count=len(ordered),
        mean_sum=sum(sums) / len(sums),
        sum_range_80=(lo, hi),
        odd_distribution=odd_d,
        low_distribution=low_d,
        ac_distribution=ac_d,
        end_sum_mean=sum(ends) / len(ends),
        consecutive_ratio=with_consecutive / len(ordered),
        carryover_distribution=carry_d,
    )
